pkg/ruler/rule_concurrency.go

```diff
@@ -102,6 +103,13 @@ type MultiTenantConcurrencyController struct {
 	tenantConcurrency    map[string]*DynamicSemaphore
 }
 
+// RemoveTenant removes a tenant from the concurrency controller.
+func (c *MultiTenantConcurrencyController) RemoveTenant(userID string) {
+	c.tenantConcurrencyMtx.Lock()
+	defer c.tenantConcurrencyMtx.Unlock()
+	delete(c.tenantConcurrency, userID)
```

Comment:
If there's a race condition and we remove a tenant when the semaphore is != 0 then we'll introduce an inconsistency and possibly a panic once `Done()` will be called. I would remove a tenant only if `acquired` is `0` and return a `bool` from `RemoveTenant()` to signal whether the tenant was removed or not (if the tenant doesn't exist it should return `true`), so that the caller can log it as an error if tenant was not removed.